### DISCUSSION TITLE
Fiori for iOS migration to SwiftUI

### DISCUSSION THREAD

**Discussion by MarcoEidinger** [SAP/cloud-sdk-ios-fiori#183] [Show and tell]
[SAP Cloud Platform SDK for iOS](https://developers.sap.com/topics/cloud-platform-sdk-for-ios.html) offers UIKit-based implementation of [Fiori for iOS](https://experience.sap.com/fiori-design-ios/) with `SAPFiori` framework, a closed-code binary framework of the SDK.

In light of our open source activities and the acknowledgment of the importance of SwiftUI, we are proud to announce the migration plans of "Fiori for iOS" to SwiftUI.

- In Q4 2020 we evaluated different ways for component API definitions and ways to ensure a consistent generation of those components.
- In January 2021 we are planning to start development with SDK developers (maintainers) in this repo
- Around Apple's WWDC 2021 event (June) we are planning to complete migration for convertible elements (and ad-hoc wrapping of UIKit-only components)

Community members can actively participate by providing early feedback! We introduced branch [migration](https://github.com/SAP/cloud-sdk-ios-fiori/tree/migration) for the time being.

**Comment by MakeSAPGreatAgain** [comment]
Hello Marco,
I am a little bit confused, transition to swiftui hasn't already been happened ?
I just thought only the sample application the assistant produce needed to update to use swiftui.

**Reply by MarcoEidinger** [reply]
[SAPFiori](https://help.sap.com/doc/978e4f6c968c4cc5a30f9d324aa4b1d7/Latest/en-US/Documents/Frameworks/SAPFiori/index.html) has a rich number of UI components that implement the Fiori Design Language. These controls inherit from Apple UIKit-based components. It is possible to wrap those components with `UIViewRepresentable` and use them in a SwiftUI application but we'd like to offer developers a better experience when developing with SwiftUI.

This repository already contains a small set of native SwiftUI controls (i.e. charts). So you are correct by saying that the transition to SwiftUI began. But components like Object/KPI Header or Contact Cell (just to name a few) are not yet available as native SwiftUI controls.

This announcement is to emphasize our planned efforts to provide native SwiftUI controls on a large scale.

As a final step, the assistant app may then use those native SwiftUI controls to generate an application with lifecycle `SwiftUI App`

**Reply by MakeSAPGreatAgain** [reply]
Thanks for the clarification, i don't know how i get the impression full native SwiftUI controls buried deep in the latest sdk  and only think left is to update lifecycle , sample application in the assistant. 

While this means we have to wait little longer , good news is , it is on the horizon and will be open source.
Thanks in advance.

What would be the best practice, your advice,  for a person who has to start a project right now and don't want to use storyboard way also be ready for the future swiftui version of the sdk  for a smooth update.

**Comment by MarcoEidinger** [comment]
We just released [1.0.0](https://github.com/SAP/cloud-sdk-ios-fiori/releases/tag/1.0.0) and the following SwiftUI components are available

| | FioriSwiftUICore |
| - | - |
| ObjectItem | :white_check_mark: | 
| ObjectHeader | :white_check_mark: | 
| KPIItem | :white_check_mark: | 
| FioriButton | :white_check_mark: |
| ListPickerItem | :white_check_mark: |
| DimensionSelector | :white_check_mark: |
| SideBar | :white_check_mark: |
| DataTable | :white_check_mark: |
| WelcomeScreen | :white_check_mark: |
| ActivationScreen | :white_check_mark: |
| InfoView | :white_check_mark: |
| KPIProgressItem | :soon: | 
| EULAView | :soon: |
| ContactItem | :soon: | 
| KeyValueItem | :soon: |
| TimelineItem | :x: |
| TimelinePreviewItem | :x: |
| ChartFloorplan | :x: |
| ProfileHeader | :x: |
| CollectionItem | :x: |
| ProfileHeader | :x: |
| BarcodeScanner | :x: |
| PasscodeView | :x: |
| MultiUserPasscodeView | :x: |
| TouchIDErrorView | :x: |
